docs/modules/ROOT/pages/references/architecture/metering-data-flow.adoc

```diff
@@ -0,0 +1,527 @@
+This page explains how that data gets transformed so that it can be ingested into a billing/ERP system.
+====
+
+The core of this part of the system has three responsibilites.
+First it must fetch the required data from the input system.
+Second that data must be enriched.
+That enrichment involves looking up prices and discounts from price tables.
+It also includes matching the identifiers from the source (Prometheus) with the identifiers from the target system.
+Third it must ingest that data into the target system.
+
+Key features of this component are:
+
+* It uses the Prometheus Query API as its source.
+* It allows to integrate any billing/ERP system as its target.
+  In order to achieve this, it must be loosely coupled.
+* Execution is idempotent.
+  If run twice for a given time frame with the same configuration, the result is the same.
+  However, if run again with a different configuration, the results reflect the changes made.
+  This ensures that errors in the past can easily be fixed.
+* Time aware configuration
+  Every configuration with the power to change the output, is made time aware.
+  This means that this config has timestamps attached denoting from when and until when that config is valid.
+
+== System idea
+
+This is an https://en.wikipedia.org/wiki/Extract,_transform,_load[ETL process].
+The extract and load part are well isolated and have only a dependency to their source/target system.
+The transform part on the other hand needs to bring parts of both the source and the target togehter.
+Instead of having one single ETL process, a multi stage model with an intermediate persistence will be used.
+The first stage fetches the data from the source into that intermediate storage.
+The second stage enriches that data by bringing in missing pieces from the target system.
+The third stage then ingests the data into the target system.
+This is done to keep this loosely coupled and thus simplify writing and maintaining the code base.
+It also allows to easily replace the code that talks to the target system.
+We take into account ease of replacing for code that talks to the target system to accommodate replacing the billing/ERP system and reselling of {product}.
+
+Many aspects of how this system works are expected to change over time.
+For example, the price for a given product and any discounts are valid only for a defined time window.
+The system deals with them by assigning timestamps to those data records.
+The timestamp `from` denotes the earliest point in time that records becomes valid.
+The timestamp `until` denotes the lates point in time that record was valid.
+If either is left blank, that record is unbound in that direction.
+So a record can be made valid from the beginning up until a certain time, while an other becomes valid up until further notice from that time on.
+The system needs to implement constraints and validations to ensure that no overlapping time ranges can be defined.
```

Comment:
It also needs to make sure if they are modified the affected facts are recomputed. The same needs to happen if we add discounts or products. This also means the whole matching logic below needs to be reapplied.

Reply:
> Do we really have the performance requirements that we need to pre-compute this

I did some tests and yes, we do have the requirements. See #83 for details.

Reply:
> I would argue in data extraction and data ingestion are quite tightly coupled by communicating through a database.

Coupling can be discussed based different aspects. I am more concerned about the coupling in the business logig. Splitting it up, allows to reduce the complexity of each individual piece. By my understanding, this makes them easier to write, test and maintain.

As for the coupling though the database, I see this as a form of API. I have concidered to have this wrapped in a HTTP or similar API. However that would be a lot of additional work with questionable benefit.